src/govuk/components/exit-this-page/exit-this-page.mjs

```diff
@@ -0,0 +1,343 @@
+  this.$updateSpan.className = 'govuk-visually-hidden'
+
+  this.$module.appendChild(this.$updateSpan)
+}
+
+/**
+ * Create button click handlers.
+ */
+ExitThisPage.prototype.initButtonClickHandler = function () {
+  // Main EtP button
+  this.$button.addEventListener('click', this.exitPage.bind(this))
+
+  // EtP skiplink
+  if (this.$skiplinkButton) {
+    this.$skiplinkButton.addEventListener('click', this.exitPage.bind(this))
+  }
+}
+
+/**
+ * Create the HTML for the 'three lights' indicator on the button.
+ */
+ExitThisPage.prototype.buildIndicator = function () {
+  // Build container
+  // Putting `aria-hidden` on it as it won't contain any readable information
+  this.$indicatorContainer = document.createElement('div')
+  this.$indicatorContainer.className = 'govuk-exit-this-page__indicator'
+  this.$indicatorContainer.setAttribute('aria-hidden', 'true')
+
+  // Create three 'lights' and place them within the container
+  for (var i = 0; i < 3; i++) {
+    var $indicator = document.createElement('div')
+    $indicator.className = 'govuk-exit-this-page__indicator-light'
+    this.$indicatorContainer.appendChild($indicator)
+  }
+
+  // Append it all to the module
+  this.$button.appendChild(this.$indicatorContainer)
+}
+
+/**
+ * Update whether the lights are visible and which ones are lit up depending on
+ * the value of `keypressCounter`.
+ */
+ExitThisPage.prototype.updateIndicator = function () {
+  // Show or hide the indicator container depending on keypressCounter value
+  if (this.keypressCounter > 0) {
+    this.$indicatorContainer.classList.add('govuk-exit-this-page__indicator--visible')
+  } else {
+    this.$indicatorContainer.classList.remove('govuk-exit-this-page__indicator--visible')
+  }
+
+  // Turn out all the lights
+  var $lightsOn = this.$indicatorContainer.querySelectorAll('.govuk-exit-this-page__indicator-light--on')
+  nodeListForEach($lightsOn, function ($light) {
+    $light.classList.remove('govuk-exit-this-page__indicator-light--on')
+  })
+
+  // Turn on the ones we want on
+  var $lightsQueried = this.$indicatorContainer.querySelectorAll('.govuk-exit-this-page__indicator-light')
+  nodeListForEach($lightsQueried, function ($light, index) {
+    if (index < this.keypressCounter) {
+      $light.classList.add('govuk-exit-this-page__indicator-light--on')
+    }
+  }.bind(this))
+}
+
+/**
+ * Initiates the redirection away from the current page.
+ * Includes the loading overlay functionality, which covers the current page with a
+ * white overlay so that the contents are not visible during the loading
+ * process. This is particularly important on slow network connections.
+ *
+ * @param {MouseEvent} [event] - mouse click event
+ */
+ExitThisPage.prototype.exitPage = function (event) {
+  var redirectUrl = this.$button.getAttribute('href')
+
+  if (typeof event !== 'undefined' && event.target instanceof HTMLElement) {
+    event.preventDefault()
+    redirectUrl = event.target.getAttribute('href')
+  }
```

Comment:
I don't understand what this is doing – can you explain?

Reply:
This is to handle the different ways that the button can be activated. If the button was clicked, we `preventDefault` and do some extra stuff before navigating. If it was activated via the keypress, we don't want to `preventDefault` as the keypress activation doesn't pass a `MouseEvent` to the the function.

We probably don't need to separate the ways we define `redirectUrl` here but we do need to have the `preventDefault` check. I'll add a clarifying comment.

Reply:
Gotcha, so `exitPage` can be called either:

- directly as an event handler for the `click` event
- indirectly from the event handler for the `keypress` event (`handleKeypress`)

Have you considered adding a `handleClick` function as an intermediary?

```js
ExitThisPage.prototype.handleClick = function (event) {
    event.preventDefault()
    this.exitPage()
}
```

That'd make the purpose of the `exitPage` method clearer and mean that you never need to pass it the `event` (which at the minute is optional).